src/peer/socket.rs

```diff
@@ -99,103 +91,107 @@ async fn retainer<S: Signer>(
-                    log::trace!("sending message to relay {:?}", &u.domain());
-                    if let Err(err) = write.send(message).await {
-                        // probably connection closed as well, we need to renew!
-                        log::error!("disconnected: error while sending message: {}", err);
-                        break 'receive;
+                read_err = closed.recv() => {
+                    if let Some(err) = read_err {
+                        log::error!("read error: {}", err);
                     }
+                    log::debug!("read routine exited, reconnecting");
+                    // force reconnecting
+                    break 'receive;
                 },
-                message = read.recv() => {
-                    let message = match message {
-                        None=> {
-                            log::error!("disconnected: read stream ended")  ;
-                            break 'receive;
-                        },
-                        Some(message) => message,
-                    };
+                message = timeout(PING_INTERVAL, up_rx.recv()) => {
+                    // first check if we have timed out
+                    let now = timestamp();
+                    if now - ts.load(Ordering::Relaxed) > READ_TIMEOUT {
+                        // we have timed out! we need to reconnect then
+                        log::info!("connection timeout! retrying");
+                        break 'receive
+                    }
 
-                    // we take a note with when a message was received
-                    last = Instant::now();
-                    log::trace!("received a message from relay {:?}", &u.domain());
                     let message = match message {
-                        Ok(Message::Pong(_)) => {
-                            log::trace!("received a pong");
-                            continue 'receive;
-                        }
-                        Ok(message) => message,
-                        Err(err) => {
-                            // todo: those errors probably mean we need to re-connect
-                            // we will see what to do later.
-                            log::error!("disconnected: error while receiving message: {}", err);
-                            break 'receive;
+                        Ok(Some(message)) => message,
+                        Ok(None) => {
+                            //weird why would we receive a non message
+                            log::error!("socket closed!");
+                            handler.abort();
+                            return
+                        },
+                        Err(_) => {
+                            // receive timeout (on upstream message)
+                            // we can then send a ping to keep the connection alive
+                            log::debug!("sending a ping");
+                            Message::Ping(Vec::default())
```

Comment:
q: why we only sending Ping on err? because this handled by the lib?

Reply:
We are not sending ping on error, we are sending pin on timeout. Wrapping the `recv()` in a timeout() makes the return of recv() wrapped in another Result. That result can be error if the timeout happened before the recv. which means it has been some time (20 seconds) with no activity and we then can send a ping 

In either cases if there is no received messages in a window or 40 seconds (2 pings) we assume the connection is dead.

Reply:
still didn't wrap my head around it.
should we then store the ts anytime we receive a message (instead of only on receving pong), so we do ping only on idle connection (when we are not receiving messages from relay)